### DISCUSSION TITLE
Next.js link prefetch causes scrolling to top

### DISCUSSION THREAD

**Discussion by Ma-Kas** [vercel/next.js#68265] [Help]
### Summary

Strange issue. I have a page to view a blog post, that is fetched from the cms. At the very bottom of the page I have a footer section for the blog post with its categories, tags, and related posts. Those are all link components to the respective pages.

When scrolling down the blog post, at some point the scroll just jumps back to the top of the page. Managed to figure out through DevTools it happens the moment any of those link components come into view and get prefetched.

Not really sure where to even start, as prefetching links shouldn't cause scrolling, and doesn't on any other page. Happens only the first time. Once the fetch is cached, it's all good. The other strange thing, it doesn't happen locally, only after deploying to Vercel, so a bit difficult to debug. I could replace Links with anchor tags or make a wrapper component with prefetch disabled, but would like the prefetch.

Code of the component (nothing special to see though): https://gist.github.com/Ma-Kas/e30a5d0aeae51ee4c2a3055dd8e81182

**Comment by hibikiledo** [comment]
I encountered this weird issue as well. I'm sharing my specific workaround in cases it helps someone.

From nextjs source code, the `InnerScrollAndFocusHandler` component which in my case causes the scrolling issue. It uses `findDOMNode()` to find a node that:
- Is an instance of `HTMLElement`
- Must not be positioned sticky or fixed 
- Must not be hidden

[Ref](https://github.com/vercel/next.js/blob/v14.2.26/packages/next/src/client/components/layout-router.tsx#L223)

When debugging, I found that `findDOMNode()` sometimes returns a node inside the `<head>` element instead of what it should be (a node that wraps my content inside `<body>`). When this happens, the subsequent instructions will fail to find a node that satisfies the requirements above and return early [here](https://github.com/vercel/next.js/blob/v14.2.26/packages/next/src/client/components/layout-router.tsx#L226). **This caused the scroll task to be pending (focusAndScrollRef.apply stays `true`)**

Later on, when a link with `prefetch` enabled comes into view, the `InnerScrollAndFocusHandler` component is re-rendered due to the updated context value. The routine that handles the scrolling then runs again. This causes the scroll position to be 0 (scroll to the top).

#### The workaround

I went through the component tree in my codebase and found that there is a client component that fetches remote data. While it is fetching, it returns `null`. This causes `findDOMNode()` to return an incorrect node (`<head>` element).

The workaround is to make sure that `findDOMNode()` can always find a node that satisfies the requirements. In my case, I replaced `null` with an empty `<div />`.
